File: src/com/machfour/macros/core/Meal.java
package com.machfour.macros.core;

import com.machfour.macros.data.ColumnData;
import com.machfour.macros.data.Columns;
import com.machfour.macros.data.Table;
import com.machfour.macros.data.Tables;
import com.machfour.macros.util.DateStamp;
import com.sun.istack.internal.NotNull;
import com.sun.istack.internal.Nullable;

import java.util.ArrayList;
import java.util.List;

public class Meal extends MacrosEntity<Meal> {

    //public static final MacrosPersistable.Converter<MealTable> CONVERTER = Converter.getInstance();
    private final List<FoodPortion> foodPortions;

    public Meal(ColumnData<Meal> data, boolean isFromDb) {
        super(data, isFromDb);
        foodPortions = new ArrayList<>();
    }

    public static NutritionData sumNutritionTotals(List<Meal> meals) {
        List<NutritionData> nutritionTotals = new ArrayList<>(meals.size());
        for (Meal m : meals) {
            nutritionTotals.add(m.getNutritionTotal());
        }
        return NutritionData.sum(nutritionTotals);
    }

    @Override
    public Table<Meal> getTable() {
        return Tables.MealTable.getInstance();
    }

    public NutritionData getNutritionTotal() {
        List<NutritionData> nutritionComponents = new ArrayList<>(foodPortions.size());
        for (FoodPortion fp : foodPortions) {
            nutritionComponents.add(fp.getNutritionData());
        }
        return NutritionData.sum(nutritionComponents);
    }

    @Override
    public boolean equals(Object o) {
        return o instanceof Meal && super.equals(o);
    }

    @Nullable
    public String getDescription() {
        return getTypedDataForColumn(Columns.MealCol.DESCRIPTION);
    }

    public DateStamp getDay() {
        return getTypedDataForColumn(Columns.MealCol.DAY);
    }

    public List<FoodPortion> getFoodPortions() {
        return new ArrayList<>(foodPortions);
    }

    public void addFoodPortion(@NotNull FoodPortion fp) {
        assert (getId().equals(fp.getMealId()));
        assert (equals(fp.getMeal()));
        assert (!foodPortions.contains(fp));
        foodPortions.add(fp);
    }
}

File: src/main/java/com/machfour/macros/entities/Meal.kt
@file:Suppress("EqualsOrHashCode")

package com.machfour.macros.entities

import com.machfour.macros.core.MacrosEntityImpl
import com.machfour.macros.entities.auxiliary.Factories
import com.machfour.macros.nutrientdata.FoodNutrientData
import com.machfour.macros.orm.Factory
import com.machfour.macros.orm.ObjectSource
import com.machfour.macros.orm.schema.FoodPortionTable
import com.machfour.macros.orm.schema.MealTable
import com.machfour.macros.sql.RowData
import com.machfour.macros.sql.Table
import com.machfour.macros.util.DateStamp
import java.time.Instant

class Meal internal constructor(data: RowData<Meal>, objectSource: ObjectSource) : MacrosEntityImpl<Meal>(data, objectSource) {
    companion object {
        fun sumNutrientData(meals: Collection<Meal>): FoodNutrientData {
            return FoodNutrientData.sum(meals.map { it.nutrientTotal() })
        }

        // factory before table
        val factory: Factory<Meal>
            get() = Factories.meal
        val table: Table<Meal>
            get() = MealTable.instance
    }

    private val foodPortions: MutableList<FoodPortion> = ArrayList()

    override val factory: Factory<Meal>
        get() = Companion.factory

    override val table: Table<Meal>
        get() = Companion.table

    fun nutrientTotal(): FoodNutrientData {
        val allNutrientData = foodPortions.map { it.nutrientData }
        return FoodNutrientData.sum(allNutrientData, null)
    }

    val name: String
        get() = getData(MealTable.NAME)!!

    /*
     * 'Day' is the day for which the nutrition data should be counted.
     * i.e. the total nutrition value for day X is the sum of nutrition data
     * for all meals having that Day field
     */
    val day: DateStamp
        get() = getData(MealTable.DAY)!!

    /*
     * Start time is the time that the meal was actually consumed. Note that
     * it's a full timestamp - because for various reasons (timezones, night shifts, eating past midnight)
     * we allow the day that the meal is labelled with in the Day column to be different from
     * the calendar date on which it was actually eaten.
     */
    // returns time in Unix time, aka seconds since Jan 1 1970
    val startTime: Long
        get() = getData(MealTable.START_TIME)!!

    // in seconds, how long the meal lasted.
    val durationSeconds: Int
        get() = getData(MealTable.DURATION)!!

    val durationMinutes: Int
        get() = durationSeconds / 60

    val startTimeInstant: Instant
        get() = Instant.ofEpochSecond(startTime)

    val notes: String?
        get() = getData(MealTable.NOTES)

    override fun equals(other: Any?): Boolean {
        return other is Meal && super.equals(other)
    }

    fun getFoodPortions(): List<FoodPortion> {
        return ArrayList(foodPortions)
    }

    fun addFoodPortion(fp: FoodPortion) {
        // can't assert !foodPortions.contains(fp) since user-created food portions can look identical
        require(foreignKeyMatches(fp, FoodPortionTable.MEAL_ID, this))
        foodPortions.add(fp)
    }

    internal fun removeFoodPortion(fp: FoodPortion) {
        val index = foodPortions.indexOf(fp)
        require(index != -1) {
            "removeFoodPortion(): FoodPortion (id = ${fp.id}) not found in meal $name (id = $id)" }
        foodPortions.removeAt(index)
    }

    override fun toString(): String = "$name, $day"
}
